File: face/geomajas-face-dojo/dojo-client/src/main/resources/geomajas/spatial/geometry/LineString.js
/*
 * This file is part of Geomajas, a component framework for building
 * rich Internet applications (RIA) with sophisticated capabilities for the
 * display, analysis and management of geographic information.
 * It is a building block that allows developers to add maps
 * and other geographic data capabilities to their web applications.
 *
 * Copyright 2008-2010 Geosparc, http://www.geosparc.com, Belgium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

dojo.provide("geomajas.spatial.geometry.LineString");
dojo.declare("LineString", Geometry, {

	/**
	 * @fileoverview Geometry implementation for LineStrings.
	 * @class Geometry implementation for LineStrings.
	 * @author Pieter De Graef
	 *
	 * @constructor
	 * @extends Geometry
	 */
	constructor : function () {
		this.coordinates = [];
	},

	/**
	 * Create a copy of this geometry and return it.
	 */
	clone : function () {
		var fac = this.getGeometryFactory();
		var coords = [];
		for(var i=0; i<this.coordinates.length; i++) {
			coords[i] = this.coordinates[i].clone();
		}
		return fac.createLineString(coords);
	},

	/**
	 * Return a coordinate, or null. 
	 * @param n Index in the geometry. This can be an integer value or an array
	 *          of values.
	 * @return A coordinate or null.
	 */
	getCoordinateN : function (n) {
		return this.getGeometryN(n);
	},

	/**
	 * Returns a coordinate, or the geometry itself.
	 * @param n Index in the geometry. This can be an integer value or an array
	 *          of values. Tries to get the coordinate n or n[0]. If that fails
	 *          it will return this linestring object. 
	 * @return A geometry object.
	 */
	getGeometryN : function (n) {
		if (n instanceof Array) {
			if (n.length == 0) {
				return this;
			} else if (n[0] >= 0 && n[0] < this.coordinates.length) {
				return this.coordinates[n[0]];
			}
		} else if (n >= 0 && n<this.coordinates.length) {
			return this.coordinates[n];
		}
		return this;
	},

	/**
	 * Returns the number of coordinates.
	 */
	getNumGeometries : function () {
		return this.getNumPoints();
	},

	/**
	 * Add a coordinate to the list. Shouldn't be here!
	 */
	appendCoordinate : function (coordinate) {
		var linestring = this._getPrimaryCoordinates();
		var length = linestring.length;
		linestring[length] = coordinate;
	},

	/**
	 * Returns 0.
	 */
	getArea : function () {
		return 0;
	},

	/**
	 * Returns the length of the linestring.
	 */
	getLength : function () {
		var len = 0;
		for (var i=0; i<this.coordinates.length-1; i++) {
			var deltaX = this.coordinates[i+1].getX() - this.coordinates[i].getX();
			var deltaY = this.coordinates[i+1].getY() - this.coordinates[i].getY();
			len += Math.sqrt(deltaX*deltaX + deltaY*deltaY);
		}
		return len;
	},

	/**
	 * The centroid is also known as the "centre of gravity" or the
	 * "center of mass".
	 * @return Return the centre point.
	 */
	getCentroid : function () {
		if (this.coordinates.length % 2 == 1) {
			var len = Math.floor(this.coordinates.length / 2);
			return this.coordinates[len];
		} else {
			var i2 = this.coordinates.length / 2;
			var i1 = i2-1;
			var ls = new LineSegment(this.coordinates[i1], this.coordinates[i2]);
			return ls.getMiddlePoint();
		}
	},

	
	/**
	 * Returns the minimal distance between this coordinate and any line segment of the geometry
	 * @return Return the minimal distance
	 */
	
	getDistance : function (coordinate) {
		var mathLib = new MathLib();
		var minDistance = Number.MAX_VALUE;

		for (var i=0; i<this.coordinates.length-1; i++) {
			var dist = mathLib.distance(this.coordinates[i], this.coordinates[i+1],coordinate);
			if (dist  < minDistance){
				minDistance = dist;
			}
		}
		return minDistance;		
	}, 
	
	/**
	 * Returns a measure for the distance between this coordinate and any line segment of the geometry
	 * @return Returns a measure for the minimal distance (approximation, not the real dist)
	 */
	
	getDistanceApprox : function (coordinate) {
		var distance = Number.MAX_VALUE;
		for (var i=0; i<this.coordinates.length; i++) {
			var dx = this.coordinates[i].getX()-coordinate.getX();
			var dy = this.coordinates[i].getY()-coordinate.getY();
			if(Math.abs(dx)+Math.abs(dy)< distance){
				distance = Math.abs(dx)+Math.abs(dy);
			}
		}
		return distance;		
	},

	/**
	 * Returns true.
	 */
	isValid : function () {
		return true;
	},

	log : function () {
		dojo.debug("LineString:");
		var coords = this.getCoordinates();
		for (var i=0; i<this.getNumPoints(); i++) {
			dojo.debug("\t"+coords[i].toString());
		}
	},

	/**
	 * @private
	 */
	_getPrimaryCoordinates : function (){
		return this.coordinates;
	}
});

File: client/geomajas-client-smartgwt/client/src/main/java/org/geomajas/smartgwt/client/spatial/geometry/LineString.java
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2013 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */

package org.geomajas.smartgwt.client.spatial.geometry;

import org.geomajas.geometry.Coordinate;
import org.geomajas.smartgwt.client.spatial.Bbox;
import org.geomajas.smartgwt.client.spatial.Mathlib;
import org.geomajas.layer.LayerType;

/**
 * LineString client-side GWT object.
 *
 * @author Pieter De Graef
 */
public class LineString extends AbstractGeometry {

	private static final long serialVersionUID = 1311848562135714815L;

	private Coordinate[] coordinates;

	// -------------------------------------------------------------------------
	// Constructor
	// -------------------------------------------------------------------------

	LineString(int srid, int precision) {
		super(srid, precision);
	}

	LineString(int srid, int precision, Coordinate[] coordinates) {
		super(srid, precision);
		this.coordinates = coordinates;
	}

	// -------------------------------------------------------------------------
	// Class specific functions:
	// -------------------------------------------------------------------------

	/**
	 * Return a coordinate, or null.
	 *
	 * @param n
	 *            Index in the geometry. This can be an integer value or an array of values.
	 * @return A coordinate or null.
	 */
	public Coordinate getCoordinateN(int n) {
		if (isEmpty()) {
			return null;
		}
		if (n >= 0 && n < coordinates.length) {
			return coordinates[n];
		}
		return null;
	}

	/**
	 * Checks whether or not the ring is closed.
	 *
	 * @return Returns true if the first coordinate equals the last coordinate.
	 */
	public boolean isClosed() {
		return !isEmpty() && getCoordinateN(0).equals(getCoordinateN(getNumPoints() - 1));
	}

	// -------------------------------------------------------------------------
	// Geometry implementation:
	// -------------------------------------------------------------------------

	/**
	 * Calculate whether or not this geometry intersects with another. (calculates simple line intersections)
	 *
	 * @param geometry
	 *            The other geometry to check for intersection.
	 * @return Returns true or false.
	 */
	public boolean intersects(Geometry geometry) {
		if (geometry == null || isEmpty() || geometry.isEmpty()) {
			return false;
		}

		if (geometry.getNumGeometries() > 1) {
			for (int n = 0; n < geometry.getNumGeometries(); n++) {
				if (intersects(geometry.getGeometryN(n))) {
					return true;
				}
			}
		} else {
			final Coordinate[] coords = geometry.getCoordinates();
			if (coords.length > 1 && coordinates.length > 1) {
				for (int i = 0; i < coordinates.length - 1; i++) {
					for (int j = 0; j < coords.length - 1; j++) {
						if (Mathlib.lineIntersects(coordinates[i], coordinates[i + 1], coords[j], coords[j + 1])) {
							return true;
						}
					}
					if (Mathlib.touches(geometry, coordinates[i])) {
						return true;
					}
				}
			} else {
				// TODO implement me
			}
		}

		return false;
	}

	public Coordinate[] getCoordinates() {
		return coordinates;
	}

	void setCoordinates(Coordinate[] coordinates) {
		this.coordinates = coordinates;
	}

	@Override
	public Object clone() { // NOSONAR super.clone() not supported by GWT
		return getGeometryFactory().createLineString(coordinates);
	}

	/**
	 * Return the geometry itself.
	 *
	 * @param n
	 *            integer value...doesn't matter.
	 *
	 * @return this
	 */
	public Geometry getGeometryN(int n) {
		return this;
	}

	public int getNumPoints() {
		if (isEmpty()) {
			return 0;
		}
		return coordinates.length;
	}

	/**
	 * Return the length of the LineString.
	 */
	public double getLength() {
		double len = 0;
		if (!isEmpty()) {
			for (int i = 0; i < coordinates.length - 1; i++) {
				double deltaX = coordinates[i + 1].getX() - coordinates[i].getX();
				double deltaY = coordinates[i + 1].getY() - coordinates[i].getY();
				len += Math.sqrt(deltaX * deltaX + deltaY * deltaY);
			}
		}
		return len;
	}

	/**
	 * The centroid is also known as the "center of gravity" or the "center of mass".
	 *
	 * @return Return the center point.
	 */
	public Coordinate getCentroid() {
		if (isEmpty()) {
			return null;
		}
		double sumX = 0;
		double sumY = 0;
		double totalLength = 0;
		for (int i = 0; i < coordinates.length - 1; i++) {
			double length = coordinates[i].distance(coordinates[i + 1]);
			totalLength += length;
			double midx = (coordinates[i].getX() + coordinates[i + 1].getX()) / 2;
			sumX += length * midx;
			double midy = (coordinates[i].getY() + coordinates[i + 1].getY()) / 2;
			sumY += length * midy;
		}
		return new Coordinate(sumX / totalLength, sumY / totalLength);
	}

	/**
	 * Return the minimal distance between this coordinate and any line segment of the geometry.
	 *
	 * @return Return the minimal distance
	 */

	public double getDistance(Coordinate coordinate) {
		double minDistance = Double.MAX_VALUE;

		if (!isEmpty()) {
			for (int i = 0; i < this.coordinates.length - 1; i++) {
				double dist = Mathlib.distance(this.coordinates[i], this.coordinates[i + 1], coordinate);
				if (dist < minDistance) {
					minDistance = dist;
				}
			}
		}
		return minDistance;
	}

	/**
	 * A linestring is valid if isEmpty() == true or coordinates.length != 1.
	 */
	public boolean isValid() {
		return isEmpty() || (coordinates.length != 1);
	}

	public boolean isEmpty() {
		return coordinates == null || coordinates.length == 0;
	}

	/**
	 * The area of a LineString is always 0.
	 */
	public double getArea() {
		return 0;
	}

	public Bbox getBounds() {
		if (isEmpty()) {
			return null;
		}
		double minX = Double.MAX_VALUE;
		double minY = Double.MAX_VALUE;
		double maxX = -Double.MAX_VALUE;
		double maxY = -Double.MAX_VALUE;

		for (Coordinate coordinate : coordinates) {
			if (coordinate.getX() < minX) {
				minX = coordinate.getX();
			}
			if (coordinate.getY() < minY) {
				minY = coordinate.getY();
			}
			if (coordinate.getX() > maxX) {
				maxX = coordinate.getX();
			}
			if (coordinate.getY() > maxY) {
				maxY = coordinate.getY();
			}
		}

		return new Bbox(minX, minY, maxX - minX, maxY - minY);
	}

	public Coordinate getCoordinate() {
		if (isEmpty()) {
			return null;
		}
		return coordinates[0];
	}

	public LayerType getLayerType() {
		return LayerType.LINESTRING;
	}

	public String toWkt() {
		if (isEmpty()) {
			return "LINESTRING EMPTY";
		}
		String wkt = "LINESTRING (";
		for (int i = 0; i < coordinates.length; i++) {
			if (i > 0) {
				wkt += ", ";
			}
			wkt += coordinates[i].getX() + " " + coordinates[i].getY();
		}
		return wkt + ")";
	}

}
